dbr quant function fusion [2/x]: use fusion for observation and inference

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#71781 dbr quant function fusion [2/x]: use fusion for observation and inference**
* #71780 fx quant: add fusion matching for operator.add and torch.relu
* #71764 dbr quant function fusion [1/x]: record matches for functions
* #71551 dbr quant: record dag of non-quantizeable ops
* #71324 dbr quant: refactor first_call hooks to be standalone
* #71312 dbr quant: rename seen_op_info to seen_q_op_info

Summary:

The previous PR added information about fusions found in the subgraphs.

This PR uses that information for:
1. inserting observers at the end of fusions and not in the middle
2. during inference, replacing the original op with the fused op. The
way this is implemented is that the base op is replaced with the fused op,
and all other ops are replaced with identity functions.

Test plan:

```
python test/test_quantization.py TestQuantizeDBR.test_fusion_functions
```

Differential Revision: [D33775097](https://our.internmc.facebook.com/intern/diff/D33775097)

vkuzo committed:
dbr quant function fusion [2/x]: use fusion for observation and inference

Summary:

The previous PR added information about fusions found in the subgraphs.

This PR uses that information for:
1. inserting observers at the end of fusions and not in the middle
2. during inference, replacing the original op with the fused op. The
way this is implemented is that the base op is replaced with the fused op,
and all other ops are replaced with identity functions.

Test plan:

```
python test/test_quantization.py TestQuantizeDBR.test_fusion_functions
```

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33775097).

vkuzo committed:
Update on "dbr quant function fusion [2/x]: use fusion for observation and inference"


Summary:

The previous PR added information about fusions found in the subgraphs.

This PR uses that information for:
1. inserting observers at the end of fusions and not in the middle
2. during inference, replacing the original op with the fused op. The
way this is implemented is that the base op is replaced with the fused op,
and all other ops are replaced with identity functions.

Test plan:

```
python test/test_quantization.py TestQuantizeDBR.test_fusion_functions
```

Differential Revision: [D33775097](https://our.internmc.facebook.com/intern/diff/D33775097)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33775097).

jerryzh168 commented on test/quantization/dbr/test_quantize_dbr.py:
```diff
@@ -432,12 +432,27 @@ def forward(self, x):
+
+        # verify that the add relu is not observed
+        self.assertTrue(
+            '1' not in mp._auto_quant_state.tensor_id_to_observer)
```
is it OK to hardcode the index here?

vkuzo replied:
yeah, it's not ideal, not sure if writing a more robust util for this is worth the effort now (it may be later)

jerryzh168 commented on torch/ao/quantization/_dbr/quantization_state.py:
```diff
@@ -872,9 +909,25 @@ def _maybe_insert_output_observers(
     ):
-        func_output_obs_type = get_func_output_obs_type(seen_q_op_info)
-        output_tensor_id = seen_q_op_info.output_tensor_infos[0].id
+        is_start_of_fusion = seen_q_op_info.fusion_info and \
+            seen_q_op_info.fusion_info.is_first_element
+        is_not_start_of_fusion = seen_q_op_info.fusion_info and \
+            not seen_q_op_info.fusion_info.is_first_element
+
+        # if we are in a fusion but not at the start, do not insert observer
+        if is_not_start_of_fusion:
+            return
 
+        # if we are in a fusion and at the start, insert observer for its end
+        if is_start_of_fusion:
+            # get the output of the end of the fusion
+            cur_seen_q_op_info = get_seen_q_op_info_of_end_of_fusion(
+                seen_q_op_info, self.idx_to_seen_q_op_infos)
+            output_tensor_id = cur_seen_q_op_info.output_tensor_infos[0].id
+        else:
+            output_tensor_id = seen_q_op_info.output_tensor_infos[0].id
```
nit: I feel restructuring the code to first check if there is fusion, then check if it is start of the fusion or not might be clearer. something like:
```
if seen_q_op_info.fusion_info:
    # in fusion
    # check if it's start or not
else:
   # not in fusion
```

jerryzh168 reviewed (approved):
lg, had an inline comment about structuring code

vkuzo committed:
Update on "dbr quant function fusion [2/x]: use fusion for observation and inference"


Summary:

The previous PR added information about fusions found in the subgraphs.

This PR uses that information for:
1. inserting observers at the end of fusions and not in the middle
2. during inference, replacing the original op with the fused op. The
way this is implemented is that the base op is replaced with the fused op,
and all other ops are replaced with identity functions.

Test plan:

```
python test/test_quantization.py TestQuantizeDBR.test_fusion_functions
```

Differential Revision: [D33775097](https://our.internmc.facebook.com/intern/diff/D33775097)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33775097).

vkuzo committed:
Update on "dbr quant function fusion [2/x]: use fusion for observation and inference"


Summary:

The previous PR added information about fusions found in the subgraphs.

This PR uses that information for:
1. inserting observers at the end of fusions and not in the middle
2. during inference, replacing the original op with the fused op. The
way this is implemented is that the base op is replaced with the fused op,
and all other ops are replaced with identity functions.

Test plan:

```
python test/test_quantization.py TestQuantizeDBR.test_fusion_functions
```

Differential Revision: [D33775097](https://our.internmc.facebook.com/intern/diff/D33775097)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33775097).
